packages/fxa-settings/src/pages/Signin/SigninTotpCode/container.tsx

```diff
@@ -0,0 +1,156 @@
+import { hardNavigate } from 'fxa-react/lib/utils';
+import {
+  AuthUiError,
+  AuthUiErrorNos,
+  AuthUiErrors,
+} from '../../../lib/auth-errors/auth-errors';
+import { currentAccount, sessionToken } from '../../../lib/cache';
+import { MozServices } from '../../../lib/types';
+import VerificationMethods from '../../../constants/verification-methods';
+import VerificationReasons from '../../../constants/verification-reasons';
+import { GraphQLError } from 'graphql';
+
+export const viewName = 'signin-totp-code';
+
+export type SigninTotpCodeContainerProps = {
+  serviceName: MozServices;
+};
+
+export const SigninTotpCodeContainer = ({
+  serviceName,
+}: SigninTotpCodeContainerProps & RouteComponentProps) => {
+  const navigate = useNavigate();
+  const location = useLocation();
+  const storedLocalAccount = currentAccount();
+
+  // TBD: A bit of a work around. In the previous signup step
+  // a new session token should have been generated. Setting
+  // it here to ensure that the authorization header is set
+  // when it comes time to verify the totp code.
+  if (storedLocalAccount?.sessionToken) {
+    sessionToken(storedLocalAccount?.sessionToken);
+  }
+
+  const { queryParamModel } = useValidatedQueryParams(SigninQueryParams);
+  const [verifyTotpCode] = useMutation(gql`
+    mutation VerifyTotp($input: VerifyTotpInput!) {
+      verifyTotp(input: $input) {
+        success
+      }
+    }
+  `);
+
+  const handleNavigation = () => {
+    // Note: `redirectTo` looks similar to `redirectPathname`, but they
+    // work slightly differently: `redirectTo` redirects automatically,
+    // while `redirectPathname` redirects after a form submission.
+
+    if (queryParamModel.redirectToUrl) {
+      hardNavigate(queryParamModel.redirectToUrl);
+    } else if (queryParamModel.redirectPathname) {
+      navigate(queryParamModel.redirectPathname + location.search);
+    } else if (
+      storedLocalAccount?.verificationReason ===
+      VerificationReasons.CHANGE_PASSWORD
+    ) {
+      navigate(`/post_verify/password/force_password_change${location.search}`);
+    } else {
+      navigate('/settings');
+    }
+  };
+
+  const submitTotpCode = async (code: string) => {
+    try {
+      const result = await verifyTotpCode({
+        variables: {
+          input: {
+            code,
+            service: queryParamModel.service,
+          },
+        },
+      });
+      GleanMetrics.loginConfirmation.submit();
+
+      // Check authentication code
+      if (result.data?.verifyTotp.success === true) {
+        logViewEvent('flow', `${viewName}.submit`);
+        return true;
+      }
+
+      return false;
+    } catch (error) {
+      const gqlError = processGraphQLError(error);
+      return gqlError.error;
+    }
+  };
+
+  if (
+    !storedLocalAccount ||
+    !storedLocalAccount.sessionToken ||
+    storedLocalAccount.verificationMethod !== VerificationMethods.TOTP_2FA
+  ) {
+    hardNavigate(`/signin${location.search}`);
```

Comment:
I believe this should be a normal `navigate` since this page is going out with other signin pages.

Looks like we should only use `hardNavigate` for redirecting back to RPs? That's what the comment above it says anyway, heh.